selfdrive/thermald/power_monitoring.py

```diff
@@ -0,0 +1,121 @@
+import time
+import datetime
+import threading
+import random
+from statistics import mean
+from cereal import log
+
+PANDA_OUTPUT_VOLTAGE = 5.28
+
+# Helpers
+def _read_param(path, parser, default=0):
+  try:
+    with open(path) as f:
+      return parser(f.read())
+  except FileNotFoundError:
+    return default
+
+def panda_current_to_actual_current(panda_current):
+  # From white/grey panda schematic
+  return (3.3 - (panda_current * 3.3 / 4096)) / 8.25
+
+# Parameters
+def get_battery_capacity():
+  return _read_param("/sys/class/power_supply/battery/capacity", int)
+
+def get_battery_status():
+  # This does not correspond with actual charging or not. 
+  # If a USB cable is plugged in, it responds with 'Charging', even when charging is disabled
+  return _read_param("/sys/class/power_supply/battery/status", lambda x: x.strip(), '')
+
+def get_battery_current():
+  return _read_param("/sys/class/power_supply/battery/current_now", int)
+
+def get_battery_voltage():
+  return _read_param("/sys/class/power_supply/battery/voltage_now", int)
+
+def get_usb_present():
+  return _read_param("/sys/class/power_supply/usb/present", lambda x: bool(int(x)), False)
+
+def get_battery_charging():
+  # This does correspond with actually charging
+  return _read_param("/sys/class/power_supply/battery/charge_type", lambda x: x.strip() != "N/A", False)
+
+def set_battery_charging(on):
+  with open('/sys/class/power_supply/battery/charging_enabled', 'w') as f:
+    f.write(f"{1 if on else 0}\n")
+
+last_measurement_time = None           # Used for integration delta
+power_used_uWh = 0                    # Integrated power usage in uWh since going into offroad
+next_pulsed_measurement_time = None
+
+# Calculation tick
+def pm_calculate(health):
+  global power_used_uWh, last_measurement_time, next_pulsed_measurement_time
```

Comment:
Why not turn this into a class so you don't need globals?

Reply:
Could do that. Will refactor and work on this more tomorrow